nntrainer/utils/base_properties.h

```diff
@@ -133,28 +160,44 @@ template <typename T> class Property {
    *
    * @return T value
    */
-  operator T &() { return value; }
+  operator T &() { return get(); }
 
   /**
    * @brief cast operator for property
    *
    * @return T value
    */
-  operator const T &() const { return value; }
+  operator const T &() const { return get(); }
 
   /**
    * @brief get the underlying data
    *
    * @return const T& data
    */
-  const T &get() const { return value; }
+  const T &get() const {
+    NNTR_THROW_IF(value == nullptr, std::invalid_argument)
+      << "Cannot get property, property is empty";
+    return *value;
+  }
 
   /**
    * @brief get the underlying data
    *
    * @return T& data
    */
-  T &get() { return value; }
+  T &get() {
+    NNTR_THROW_IF(value == nullptr, std::invalid_argument)
+      << "Cannot get property, property is empty";
+    return *value;
```

Comment:
Your commit message, ```This patch adds empty concept to property instead of having invalid
value as a default.``` suggests that ```value == nullptr``` is allowed. Is it a code-bug? or an insufficient commit message?

Reply:
before this patch,
```cpp
auto a = Property<int>();
EXPECT_EQ(a.get(), -1); // let's assume -1 is default value which might not be sensible.
```
after this patch

```cpp
auto a = Property<int>();
EXPECT_THROW(a.get(), std::invalid_argument); 
EXPECT_TRUE(a.empty());
```

so value remains always null or valid.

Property wraps around a value. So, I'll change the commit msg to
```diff
- invalid value
+ invalid property
```

and add more explanation :)